server/src/main/java/org/apache/druid/server/LocalQuerySegmentWalker.java

```diff
@@ -80,21 +81,28 @@ public LocalQuerySegmentWalker(
   @Override
   public <T> QueryRunner<T> getQueryRunnerForIntervals(final Query<T> query, final Iterable<Interval> intervals)
   {
+    ExecutionVertex ev = ExecutionVertex.of(query);
     final DataSource dataSourceFromQuery = query.getDataSource();
-    final DataSourceAnalysis analysis = query.getDataSourceAnalysis();
 
-    if (!analysis.isConcreteBased() || !dataSourceFromQuery.isGlobal()) {
+    if (!ev.canRunQueryUsingLocalWalker()) {
       throw new IAE("Cannot query dataSource locally: %s", dataSourceFromQuery);
     }
 
     // wrap in ReferenceCountingSegment, these aren't currently managed by SegmentManager so reference tracking doesn't
     // matter, but at least some or all will be in a future PR
     final Iterable<ReferenceCountingSegment> segments =
-        FunctionalIterable.create(segmentWrangler.getSegmentsForIntervals(analysis.getBaseDataSource(), intervals))
+        FunctionalIterable.create(segmentWrangler.getSegmentsForIntervals(ev.getBaseDataSource(), intervals))
                           .transform(ReferenceCountingSegment::wrapRootGenerationSegment);
 
     final AtomicLong cpuAccumulator = new AtomicLong(0L);
 
+    if (dataSourceFromQuery instanceof QueryDataSource) {
+      // is this a possible bug?  dataSourceFromQuery is QueryDataSource ; query is GBY with canPerformSubquery
+      // FIXME: let's not fix this
+      // throw DruidException.defensive("This might not work correctly");
+    }
+
+    // segmentMapFn will ignore all DS maps ass QDS will return identity
```

Comment:
Trying to think through how this should work for queries where the toolchest can perform a subquery. In core this only happens for `groupBy` with a `query` datasource containing another `groupBy`. In this situation the toolchest returns true from `canPerformSubquery`.

IMO, in this case it would make sense to do this:

- QueryDataSource throws an error from `createSegmentMapFunction` (as mentioned in another review comment)
- Prior to calling `createSegmentMapFunction`, this code, in LocalQuerySegmentWalker, strips stacks of `query` datasources that are handleable by the toolchest. The idea is that it needs to get to the processable datasource that is embedded under the stack of toolchest-handleable `query`.

Reply:
yes; throwing an exception in that method is a great idea - I guess I should have done that earlier.



> strips stacks of query datasources that are handleable by the toolchest

I guess nothing unhandlable should arrive here.... but iirc there were something failing..
I have a feeling that this will mandate the other change I'm hesitating to make with `QueryDataSource#isGlobal` :)
I don't remember which test was affected by this - so I'll do a CI run :)